ports-of-call/span.hpp

```diff
@@ -0,0 +1,464 @@
+#ifndef PORTS_OF_CALL_SPAN_HH_
+#define PORTS_OF_CALL_SPAN_HH_
+
+// ========================================================================================
+// © (or copyright) 2019-2024. Triad National Security, LLC. All rights
+// reserved.  This program was produced under U.S. Government contract
+// 89233218CNA000001 for Los Alamos National Laboratory (LANL), which is
+// operated by Triad National Security, LLC for the U.S.  Department of
+// Energy/National Nuclear Security Administration. All rights in the
+// program are reserved by Triad National Security, LLC, and the
+// U.S. Department of Energy/National Nuclear Security
+// Administration. The Government is granted for itself and others acting
+// on its behalf a nonexclusive, paid-up, irrevocable worldwide license
+// in this material to reproduce, prepare derivative works, distribute
+// copies to the public, perform publicly and display publicly, and to
+// permit others to do so.
+// ========================================================================================
+
+#include <array>
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <cstdio>
+#include <type_traits>
+
+#if defined(__cpp_lib_span) // do we already have std::span?
+
+#include <span>
+namespace PortsOfCall::span {
+
+using std::span;
+
+} // namespace PortsOfCall::span
+
+#else
+
+#define span_REQUIRES(...) typename std::enable_if<((__VA_ARGS__)), int>::type = 0
+
+#define span_EXPECTS(...) assert((__VA_ARGS__))
+
+namespace PortsOfCall::span {
+
+using std::byte;
+using std::void_t;
+
+// set dynamic_extent to maximum integer size
+constexpr std::size_t dynamic_extent = SIZE_MAX;
+
+// forward declare
+template <class ElementType, std::size_t Extent = dynamic_extent>
+class span;
+
+namespace detail {
+
+using std::data;
+using std::size;
+
+// object to handle storage of span
+template <class E, std::size_t S>
+struct span_storage {
+  constexpr span_storage() noexcept = default;
+
+  constexpr span_storage(E *p_ptr, std::size_t /*unused*/) noexcept : ptr(p_ptr) {}
+
+  E *ptr = nullptr;
+  static constexpr std::size_t size = S;
+};
+
+// specialization for dynamic_extent
+template <class E>
+struct span_storage<E, dynamic_extent> {
+  constexpr span_storage() noexcept = default;
+
+  constexpr span_storage(E *p_ptr, std::size_t p_size) noexcept
+      : ptr(p_ptr), size(p_size) {}
+
+  E *ptr = nullptr;
+  std::size_t size = 0;
+};
+
+// some metaprogramming
```

Comment:
Is this stuff all in support of concepts?

Reply:
Yes. I had wanted have more verbose concepts, but it was taking too long. What is here is a minimal set of constraints that cover most cases.

Reply:
We should consider at some point either backporting concepts more fully into ports of call, or just waiting for C++20

Reply:
Based on what I was doing, I'd say it's not much worth it to try and replicate `std::concepts` and the like (or stuff that exists in the standard).

That said, I think we should start prototyping concepts for `singularity-eos` and `spiner`. What these would be should be given some thought, but it would be good to start thinking about how we want to use language-supported `concepts` and `requires` when they become available.

Reply:
agreed